mm2src/coins/coin_balance.rs

```diff
@@ -0,0 +1,452 @@
+    params: HDAccountBalanceParams,
+}
+
+#[derive(Deserialize)]
+pub struct HDAccountBalanceParams {
+    pub account_index: u32,
+    pub chain: Bip44Chain,
+    #[serde(default = "common::ten")]
+    pub limit: usize,
+    #[serde(default)]
+    pub paging_options: PagingOptionsEnum<u32>,
+}
+
+#[derive(Deserialize)]
+pub struct CheckHDAccountBalanceRequest {
+    coin: String,
+    #[serde(flatten)]
+    params: CheckHDAccountBalanceParams,
+}
+
+#[derive(Deserialize)]
+pub struct CheckHDAccountBalanceParams {
+    pub account_index: u32,
+    pub gap_limit: Option<u32>,
+}
+
+#[derive(Debug, PartialEq, Serialize)]
+pub struct HDAccountBalanceResponse {
+    pub account_index: u32,
+    pub derivation_path: RpcDerivationPath,
+    pub addresses: Vec<HDAddressBalance>,
+    pub limit: usize,
+    pub skipped: u32,
+    pub total: u32,
+    pub total_pages: usize,
+    pub paging_options: PagingOptionsEnum<u32>,
+}
+
+#[derive(Debug, PartialEq, Serialize)]
+pub struct CheckHDAccountBalanceResponse {
+    pub account_index: u32,
+    pub derivation_path: RpcDerivationPath,
+    pub new_addresses: Vec<HDAddressBalance>,
+}
+
+#[async_trait]
+pub trait EnableCoinBalanceOps {
+    async fn enable_coin_balance(&self) -> BalanceResult<EnableCoinBalance>;
+}
+
+#[async_trait]
+impl<Coin, Address, HDWallet, HDAccount, HDAddressChecker> EnableCoinBalanceOps for Coin
+where
+    Coin: AddressBalanceOps<Address = Address>
+        + CoinWithDerivationMethod<Address = Address, HDWallet = HDWallet>
+        + HDWalletBalanceOps<HDWallet = HDWallet, HDAccount = HDAccount, HDAddressChecker = HDAddressChecker>
+        + Sync,
+    Address: fmt::Display + Sync,
+    HDWallet: Sync,
+    HDAddressChecker: HDAddressBalanceChecker,
+{
+    async fn enable_coin_balance(&self) -> BalanceResult<EnableCoinBalance> {
+        match self.derivation_method() {
+            DerivationMethod::Iguana(address) => self.address_balance(address).await.map(|balance| {
+                EnableCoinBalance::Iguana(IguanaWalletBalance {
+                    address: address.to_string(),
+                    balance,
+                })
+            }),
+            DerivationMethod::HDWallet(hd_wallet) => self
+                .enable_hd_wallet_balance(hd_wallet)
+                .await
+                .map(EnableCoinBalance::HD),
+        }
+    }
+}
+
+#[async_trait]
+pub trait HDWalletBalanceOps: AddressBalanceOps {
+    type HDWallet;
+    type HDAccount;
+    type HDAddressChecker: HDAddressBalanceChecker<Address = Self::Address>;
+
+    async fn produce_hd_address_checker(&self) -> BalanceResult<Self::HDAddressChecker>;
+
+    /// Checks for addresses of every known account by using [`HDWalletBalanceOps::check_hd_account_balance`].
+    /// This method is used on coin initialization to index working addresses and to return the wallet balance to the user.
+    async fn enable_hd_wallet_balance(&self, hd_wallet: &Self::HDWallet) -> BalanceResult<HDWalletBalance>;
```

Comment:
I think `init_hd_wallet` will be a more clear name for this.

Reply:
I started naming RPC calls that spawn RPC tasks like `init_withdraw`, `init_utxo` etc, so is it ok for you to rename it to `enable_hd_wallet`?

Reply:
I renamed it to `enable_hd_wallet `. I can rename it to `init_hd_wallet` as well if you think it's better

Reply:
`enable_hd_wallet` also looks fine, thanks!